cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataEntity.java

```diff
@@ -0,0 +1,118 @@
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Entity representation for Metadata
+ */
+public class MetadataEntity {
+
+  public static final String NAMESPACE = "namespace";
+  public static final String APPLICATION = "application";
+  public static final String ARTIFACT = "artifact";
+  public static final String VERSION = "version";
+  public static final String DATASET = "dataset";
+  public static final String STREAM = "stream";
+  public static final String VIEW = "view";
+  public static final String TYPE = "type";
+  public static final String PROGRAM = "program";
+
+  private final List<KeyValue> details;
+
+  private MetadataEntity(List<KeyValue> details) {
+    this.details = new ArrayList<>(details);
+  }
+
+  /**
+   * Creates a {@link MetadataEntity} representing the given datasetName. Alternatively, you
+   * can also construct a {@link MetadataEntity} which is equivalent to DatasetId by providing the namespace using
+   * {@link MetadataEntity#ofDataset(String, String)}.
+   *
+   * @param datasetName the name of the dataset
+   * @return {@link MetadataEntity} representing the dataset name
+   */
+  public static MetadataEntity ofDataset(String datasetName) {
+    return new MetadataEntity(Collections.singletonList(new KeyValue(DATASET, datasetName)));
+  }
+
+  /**
+   * Creates a {@link MetadataEntity} representing the given datasetName in the specified namespace.
+   *
+   * @param namespace the name of the namespace
+   * @param datasetName the name of the dataset
+   * @return {@link MetadataEntity} representing the dataset name
+   */
+  public static MetadataEntity ofDataset(String namespace, String datasetName) {
+    return new MetadataEntity(Arrays.asList(new KeyValue(NAMESPACE, namespace),
+                                            new KeyValue(DATASET, datasetName)));
+  }
+
+  /**
+   * Creates a {@link MetadataEntity} representing the given namespace.
+   *
+   * @param ns the name of the namespace
+   * @return {@link MetadataEntity} representing the namespace name
+   */
+  public static MetadataEntity ofNamespace(String ns) {
+    return new MetadataEntity(Collections.singletonList(new KeyValue(NAMESPACE, ns)));
+  }
+
+  /**
+   * Returns a new {@link MetadataEntity} which consists of the given key and values following the key and values of
+   * this {@link MetadataEntity}
+   *
+   * @param key the key to be added
+   * @param value the value to be added
+   */
+  public MetadataEntity append(String key, String value) {
+    List<KeyValue> existingParts = new ArrayList<>(getKeyValues());
+    existingParts.add(new KeyValue(key, value));
+    return new MetadataEntity(existingParts);
+  }
+
+  /**
+   * @return A {@link List} of {@link KeyValue} representing the metadata entity
+   */
+  public List<KeyValue> getKeyValues() {
+    return Collections.unmodifiableList(details);
```

Comment:
You should do this in the constructor if the contract of the `details` list is not going to be modified.